addons/mail/models/mail_alias_domain.py

```diff
@@ -0,0 +1,99 @@
+    catchall_email = fields.Char('Catchall Email', compute='_compute_catchall_email')
+
+    @api.depends('bounce', 'name')
+    def _compute_bounce_email(self):
+        self.bounce_email = ''
+        for domain in self.filtered('bounce'):
+            domain.bounce_email = f'{domain.bounce}@{domain.name}'
+
+    @api.depends('catchall', 'name')
+    def _compute_catchall_email(self):
+        self.catchall_email = ''
+        for domain in self.filtered('catchall'):
+            domain.catchall_email = f'{domain.catchall}@{domain.name}'
+
+    _sql_constraints = [
+        (
+            'code_company_uniq',
+            'UNIQUE(name)',
+            'Alias domain name should be unique'
+        ),
+    ]
+
+    @api.constrains('bounce', 'catchall')
+    def _check_bounce_catchall_uniqueness(self):
+        names = self.filtered('bounce').mapped('bounce') + self.filtered('catchall').mapped('catchall')
+        if not names:
+            return
+
+        existing = self.env['mail.alias'].search(
+            [('alias_name', 'in', list(set(names)))],
+            limit=1,
+        )
+        if existing.alias_parent_model_id and existing.alias_parent_thread_id:
+            # If parent model and parent thread ID both are set, display document name also in the warning
+            document_name = self.env[existing.alias_parent_model_id.model].sudo().browse(existing.alias_parent_thread_id).display_name
+            raise exceptions.ValidationError(
+                _("'%(matching_alias_name)s' is already used by the %(document_name)s %(model_name)s. Choose another alias or change it on the other document.",
+                  matching_alias_name=existing.display_name,
+                  document_name=document_name,
+                  model_name=existing.alias_parent_model_id.name)
+                    )
+        if existing:
+            raise exceptions.ValidationError(
+                _("'%(matching_alias_name)s' is already linked with %(alias_model_name)s. Choose another alias or change it on the linked model.",
+                  matching_alias_name=existing.display_name,
+                  alias_model_name=existing.alias_model_id.name)
+            )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """ Sanitize bounce / catchall """
+        for vals in vals_list:
+            if vals.get('bounce'):
+                vals['bounce'] = self.env['mail.alias']._sanitize_alias_name(vals['bounce'])
+            if vals.get('catchall'):
+                vals['catchall'] = self.env['mail.alias']._sanitize_alias_name(vals['catchall'])
+
+        alias_domains = super().create(vals_list)
+
+        # alias domain init: populate companies at first creation
+        if alias_domains and self.search_count([]) == len(alias_domains):
+            self.env['res.company'].search(
+                [('alias_domain_id', '=', False)]
+            ).alias_domain_id = alias_domains[0].id
```

Comment:
isn't `_compute_alias_domain_id` enough ?

Reply:
I guess you would have to make the compute method on the company depend on "mail_alias_ids", which is not available as it seems it's not a real m2m after all? (see previous comment).

But I agree that it would be nice to get rid of this code.